core/src/main/java/org/elasticsearch/env/Environment.java

```diff
@@ -138,9 +145,32 @@ public Environment(Settings settings) {
                 dataFiles[i] = PathUtils.get(dataPaths.get(i));
                 dataWithClusterFiles[i] = dataFiles[i].resolve(clusterName.value());
             }
+            if (DEFAULT_PATH_DATA_SETTING.exists(settings)) {
+                final String defaultPathDataValue = DEFAULT_PATH_DATA_SETTING.get(settings);
+                final Set<Path> dataFilesSet = Arrays.stream(dataFiles).collect(Collectors.toSet());
+                final Path defaultPathData = PathUtils.get(defaultPathDataValue);
+                if (dataFilesSet.size() == 1 && dataFilesSet.contains(defaultPathData)) {
+                    // default path data was used to set path data
+                    this.defaultPathData = null;
+                } else if (dataFilesSet.contains(defaultPathData)) {
+                    final String message = String.format(
+                            Locale.ROOT,
+                            "do not include default.path.data [%s] in path.data %s",
+                            defaultPathData,
+                            Arrays.toString(dataFiles));
+                    throw new IllegalStateException(message);
+                } else {
+                    this.defaultPathData = defaultPathData;
+                }
+            } else {
+                defaultPathData = null;
+            }
         } else {
             dataFiles = new Path[]{homeFile.resolve("data")};
             dataWithClusterFiles = new Path[]{homeFile.resolve("data").resolve(clusterName.value())};
+            assert !DEFAULT_PATH_DATA_SETTING.exists(settings)
+                    : "expected default.path.data to be unset but was [" + DEFAULT_PATH_DATA_SETTING.get(settings) + "]";
```

Comment:
We are not setting `path.data` from the default, and we are not nullifying `default.path.data`. We are capturing that `default.path.data` is not set, and defaulting `dataFiles` since neither `path.data` nor `default.path.data` are set.